### PR TITLE
ATB integration tests

### PR DESCRIPTION
**Pull request by andrey-p**
**Reviewer:** @jdorweiler 

**CC:** @jkv 

## Description:

Adds integration tests to the ATB workflow (and more generally, an integration test harness).

## Steps to test this PR:
1. `npm run test-int`, tests should pass (`npm run test-int-x` if you've got a Linux box)
2. If you've got some time e.g. going off to lunch, you can run something like the following to run the tests repeatedly until one of them breaks:

```bash
$ while npm run test-int; do :; done
```

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: integration-test/background/atb.js [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        ({ browser, bgPage, requests } = await harness.setup())
+
+        // wait until normal exti workflow is done so we don't confuse atb.js requests
+        // when the actual tests run
+        await wait.forSetting(bgPage, 'extiSent')
+        await bgPage.evaluate(() => dbg.settings.updateSetting('atb', 'v112-1'))
+
+        // request needs to be promisified
+        await new Promise(resolve => {
+            request('https://duckduckgo.com/atb.js', (err, res, body) => {
```
[the remark this answers is not in the document]
Comment (andrey-p) [reply integration-test/background/atb.js]: Good call.

File: integration-test/helpers/harness.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+    const browser = await puppeteer.launch(puppeteerOps)
+    // for some reason we need to init a blank page
+    // before the extension is initialized
+    await browser.newPage()
+    const targets = await browser.targets()
+    let bgPage
+
+    // grab a handle on the background page for the extension
+    // we can't use the long ID as it could possibly change
```
[the remark this answers is not in the document]
Comment (andrey-p) [reply integration-test/helpers/harness.js]: When I tested locally, I had another background page. I suspect that may be something to do with my local machine setup, but I'd rather at least verify we've got the right background page.

File: integration-test/helpers/wait.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+/* global dbg:false */
+const ms = async (ms) => {
+    await new Promise((resolve) => { setTimeout(resolve, ms) })
+}
```
[the remark this answers is not in the document]
Comment (andrey-p) [reply integration-test/helpers/wait.js]: There is, but I think it's good to have an alternative that doesn't depend on having a `page` object lying around.

(I assume the built-in method uses something similar under the hood.)

File: integration-test/background/atb.js [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    describe('postinstall page', () => {
+        beforeEach(async () => {
+            ({ browser, bgPage, requests } = await harness.setup())
+        })
+        afterEach(async () => {
+            await harness.teardown(browser)
+        })
+
+        it('should open the postinstall page correctly', async () => {
+            await wait.ms(2000)
```
[the remark this answers is not in the document]
Comment (andrey-p) [reply integration-test/background/atb.js]: I assumed someone might pick me up on that :) I'll add a proper wait.

File: .travis.yml [hunk 5]
```diff
@@ -1,9 +1,30 @@
 language: node_js
 node_js:
   - 8.11.1
-script: npm run lint && npm test
+script: npm run lint && npm test && npm run test-int-x
```
[the remark this answers is not in the document]
Comment (andrey-p) [reply .travis.yml]: Good idea.

File: integration-test/background/atb.js [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+let requests
+
+describe('install workflow', () => {
+    describe('postinstall page', () => {
+        beforeEach(async () => {
+            ({ browser, bgPage, requests } = await harness.setup())
+        })
+        afterEach(async () => {
+            await harness.teardown(browser)
+        })
```
[the remark this answers is not in the document]
Comment (andrey-p) [reply integration-test/background/atb.js]: That's a very good question.

When I put the test together the first time, I was assuming I needed to capture all the stuff that happened immediately when the extension is installed. It was before I figured I could just do stuff like this: https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/dfd5ee4306e7d38983d9b7a0b77cc7877a4f34f6/integration-test/background/atb.js#L43-L53

Maybe the `harness` helper can create the browser reference first time, then just pass the same cached browser instance. But then we need to worry about making sure everything is clear between individual tests (e.g. tabs, extension settings, etc) which might be a faff.

**Comment by andrey-p** [comment]
Thanks for the review @kdzwinel!

> with jest we could run multiple tests in paralel

I've not tried it, but that sounds possible. I think there's a benefit in using the same testing framework across both integration and unit tests, but seeing the APIs are similar it might be easy to completely switch over at a later date.

There could be other ways of parallelizing, e.g. just spawn a separate Jasmine process for each integration test file we add.

> we can mock all external request, so that we don't have to wait for the network

One of the goals here was to make this test cover as much of the end-to-end process as possible. E.g. if our atb.js service starts spouting things that the extension doesn't understand (or vice versa), we should have something failing somewhere. (The unit tests mock out requests, of course.)

File: integration-test/background/atb.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrey-p) [reply integration-test/background/atb.js]: I might punt this for now. It wouldn't be too difficult to address at a later date.

File: Gruntfile.js [hunk 7]
```diff
@@ -213,7 +213,11 @@ module.exports = function(grunt) {
     }))
 
     grunt.registerTask('build', 'Build project(s)css, templates, js', ['sass', 'browserify:ui', 'browserify:background', 'browserify:backgroundTest', 'execute:preProcessLists', 'safari'])
-    grunt.registerTask('dev', 'Build and watch files for development', ['build', 'watch'])
+
+    let devTasks = ['build']
```
Comment (kdzwinel) [inline Gruntfile.js]: I keep seeing that we are using `let` almost exclusively (in this and all other PRs). Is there any reason for that? `const`s can save us from some nasty bugs and make it a bit easier (IMO) to read the code (by suggesting that given value will or will not be soon overwritten). Some more info: https://eslint.org/docs/rules/prefer-const

Comment (andrey-p) [reply Gruntfile.js]: This is a good question. I guess we don't have strict guidelines on `let` vs `const`.

Personally, I've tended to use `let` for objects/arrays to mean that they'll be mutated in the future, but thinking about it more, it's not ideal (and others [seem to agree](https://stackoverflow.com/questions/46553062/what-is-the-preferred-declaration-convention-for-objects-or-arrays-const-or-let)).

I'll switch to `const` here, please feel free to point out places in other PRs where I forget to do that!

**Pull request merged**